swarm/storage/localstore.go

```diff
@@ -85,6 +93,13 @@ func (self *LocalStore) CacheCounter() uint64 {
 // LocalStore is itself a chunk store
 // unsafe, in that the data is not integrity checked
 func (self *LocalStore) Put(chunk *Chunk) {
+	if self.Validator != nil {
```

Comment:
I don't like the fact that we have to check if there is a Validator or not. Organising behaviour should be done with interfaces, not with defensive coding and nil checks. The fact that a Validator can be nil, has to be checked everywhere we use a Validator now. I think we should change this and make the design so that Validator is always there upon initialisation.

Reply:
The point is that it should be possible to operate _without_ a validator. I can alias this with `HasValidator()` or something, but that also means adding rather useless stack frames.

Reply:
If this should operator without a validator, then you have an interface and a type that do nothing - it is much better than having to check if a value is present or not everywhere.

Reply:
> it is much better than having to check if a value is present or not everywhere.

But seriously, where else will you have this call?

Reply:
I am suggesting to change the design for the purpose of readability and avoiding bugs, not because of performance.

Reply:
> If this should operate without a validator, then you have an interface and a type that do nothing

It should be possible to run without content validator, without resource validator or both. The latter means run without a validator.

That means, if I understand your suggestion right, that we should have:

- ContentAndResourceValidator
- ContentValidator
- ResourceValidator
- NoopValidator

?

Reply:
Yes, this is what I am suggesting.

Probably worth splitting the validation to 2 validators, then you'd only have 3 validator, rather than all the combinations:

ContentValidator
ResourceValidator
NoopValidator

Reply:
Doesn't then the logic that's currently in `ChunkValidator.Validate` just get pushed up the stack?